resources/cpp.markdown

```diff
@@ -0,0 +1,661 @@
+constructor with the signature:
+
+```cpp
+Rectangle::Rectangle(const Rectangle& x)
+```
+
+Likewise, the *destructor* function is called each time an object of the
+class is destroyed. It has the same name as the class (and the
+constructor), but is preceded with the tilde sign (`~`). It is typically
+used to perform some cleanup of heap allocated memory.
+
+When a class has dynamic member variables, these member variables are
+default-constructed when an instance of the class is created (when the
+constructor is called), and then re-constructed when they are
+initialized. To avoid this wasteful double-construction of member
+variables, constructors can use initialization lists before their
+bodies, ensuring that the member variables (width and height below) are
+only constructed once:
+
+```cpp
+Rectangle::Rectangle(int x, int y) : width(x), height(y) { }
+```
+
+#### `this` keyword
+
+> The keyword
+> [this](http://www.cs.technion.ac.il/users/yechiel/c++-faq/using-this-in-ctors.html)
+> represents a pointer to the object whose member function is being
+> executed. It is used within a
+> [class](http://www.cplusplus.com/doc/tutorial/classes/)'s member
+> function to refer to the object itself.
+
+It is often used to resolve ambiguity between a member variable of the
+object executing the call and an object of the same class passed as a
+parameter. An example illustrating this is shown in the member function
+[isitme](http://th.cpp.sh/5m2).
+
+#### Static members
+
+A static member of a class is a variable that is shared between all
+objects of the class. Likewise, a static member function is not
+associated with any specific object. Therefore, they do not have access
+to a `this` pointer and are instead accessed by the class's name, like so:
+
+```cpp
+ClassName::staticMemberVariable
+```
+
+#### Friendship
+
+The `friend` keyword defines access relationships between functions and
+classes. If a class A is a "friend" of class B, member functions in
+class A can access private and protected members of class B. Friendship
+is defined in the class giving access (Class B in this example). An
+example of friend classes is shown [here](http://th.cpp.sh/3ws).
+
+Similarly, if a function is a "friend" of class B, it can access private
+and protected members of class B within its function body. An example of
+a friend function is shown [here](http://th.cpp.sh/3bu).
+
+#### Inheritance
+
+Inheritance creates an "*is-a*" relationship between the *derived* and
+*base* class. The derived class inherits member variables and functions
+from the base class. It can also include its own member variables and
+functions. An example can be found [here](http://th.cpp.sh/6sv7cq).
+The Rectangle class is derived from the base class, Shape.
+
+Recall that derived classes can access protected members in the base
+class.
+
+As shown in the above-linked example, inheritance is specified using
+access modifiers `public`, `protected`, or `private`:
+```cpp
```

Comment:
can you add blank lines around the code block.

Reply:
done